viewing_party/party.py

```diff
@@ -1,23 +1,197 @@
+    
+    if len(user_data["watched"]) < 1:
+        most_watched_genre = None 
+        return most_watched_genre
+
+    genre_count_dict = {}
+
+    for movie in user_data["watched"]:
+        if movie["genre"] not in genre_count_dict:
+            genre_count_dict[movie["genre"]] = 1
+        else: 
+            genre_count_dict[movie["genre"]] += 1
+
+    highest_count_sofar = 0
+    highest_genre = None
+
+    for genre, count in genre_count_dict.items():
+        if count > highest_count_sofar:
+            highest_count_sofar = count
+            highest_genre = genre
+
+    return highest_genre
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_watched_list.append(movie["title"])
+    
+    user_watched_set = set(user_watched_list)
+    friend_watched_set = set(friend_watched_list)
+    user_unique_watched_set = user_watched_set.difference(friend_watched_set)
+    user_unique_watched_list = []
+
+    for unique_movie in user_unique_watched_set:
+        for movie in user_data["watched"]:
+            if movie["title"] == unique_movie:
+                user_unique_watched_list.append(movie)
+
+    return user_unique_watched_list
+
+def get_friends_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_watched_list.append(movie["title"])
+    
+    user_watched_set = set(user_watched_list)
+    friend_watched_set = set(friend_watched_list)
+    friend_unique_watched_set = friend_watched_set.difference(user_watched_set)
+    friend_unique_watched_list = []
+
+    for unique_movie in friend_unique_watched_set:
+        for friend in user_data["friends"]:
+            for movie in friend["watched"]:
+                if movie not in friend_unique_watched_list:
+                    if movie["title"] == unique_movie:
+                        friend_unique_watched_list.append(movie)
+
+    return friend_unique_watched_list
+
+
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+
+    friends_unique_watched_list = get_friends_unique_watched(user_data)
+
+    available_recommended_movies_list = []
+
+    for movie in friends_unique_watched_list:
+        for subscription in user_data["subscriptions"]:
+            if subscription == movie["host"]:
+                available_recommended_movies_list.append(movie)

```

Comment:
Looking for the movie host in the subscriptions could be written like this:

```python
    for movie in friends_unique_watched_list:
        if movie["host"] in user_data["subscriptions"]:
            available_recommended_movies_list.append(movie)
```

This is exactly equivalent to your version, but replaces the explicit loop with an `in` check expressing the intent of seeing whether the host is in the subscriptions. The other thing this lets us think about doing is possibly calculate a `set` of the subscriptions and use that for the `in` check instead. `in` checks on sets are much more efficient than for lists, so often, if we have a list that we're making `in` checks on (especially if it's in a loop) we might consider reorganizing the list as a set for improved performance:

```python
    subscription_set = set(user_data["subscriptions"])
    for movie in friends_unique_watched_list:
        if movie["host"] in subscription_set:
            available_recommended_movies_list.append(movie)
```
This version only needs to iterate once to make the set (outside the loop) a d the inside the loop, it uses fast `in` checks.